### PR TITLE
Implement Array Iterators in javascript

### PR DESCRIPTION
**Pull request by meg-gupta**
This change adds implementation of Javascript's array iterators in Javascript.
This combined with optimization of try finally functions and enabling inlining within functions have try-catch/try-finally improves performance of es6 for-of loops.
For large for-of loops the speedup is more than 2.5x.

### CODE REVIEW

**Comment by meg-gupta** [comment]
Commits on top of "Merge conflict errors " are new code. Previous commits have been reviewed before in the builtins branch.

File: lib/Runtime/Library/JsBuiltIn/JsBuiltIn.js [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+                if (typeof (o) !== 'object') {
+                    throw new TypeError("Array Iterator.prototype.next: 'this' is not an Array Iterator object");
+                }
+
+                if (!(o instanceof __chakraLibrary.ArrayIterator)) {
+                    throw new TypeError("Array Iterator.prototype.next: 'this' is not an Array Iterator object");
+                }
+
+                let a = o.arrayObj;
+                var value, done;
```
Comment (akroshg) [inline lib/Runtime/Library/JsBuiltIn/JsBuiltIn.js]: >var value, done; [](start = 16, length = 16)

initialize this as undefined and true, then you don't need to have it set two places below?

File: lib/Runtime/Library/JsBuiltIn/JsBuiltIn.js [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+
+                if (typeof (o) !== 'object') {
+                    throw new TypeError("Array Iterator.prototype.next: 'this' is not an Array Iterator object");
+                }
+
+                if (!(o instanceof __chakraLibrary.ArrayIterator)) {
+                    throw new TypeError("Array Iterator.prototype.next: 'this' is not an Array Iterator object");
+                }
+
+                let a = o.arrayObj;
```
Comment (akroshg) [inline lib/Runtime/Library/JsBuiltIn/JsBuiltIn.js]: >arrayObj [](start = 26, length = 8)

these are enumerable and visible to the script.

**Comment by obastemur** [comment]
Looks like, there are changes related to OS9049104

File: lib/Runtime/Library/JsBuiltIn/JsBuiltIn.js [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+                var value, done;
+
+                if (a === undefined) {
+                    value = undefined;
+                    done = true;
+                }
+                else {
+                    let index = o.nextIndex;
+                    let len = Array.isArray(a) ? a.length : __chakraLibrary.GetArrayLength(a);
```
Comment (MikeHolman) [inline lib/Runtime/Library/JsBuiltIn/JsBuiltIn.js]: What happens if Array.isArray is overridden?

File: lib/Runtime/Library/JsBuiltIn/JsBuiltIn.js (hunk 1 above)
Comment (MikeHolman) [reply lib/Runtime/Library/JsBuiltIn/JsBuiltIn.js]: Nit: use let instead of var

File: lib/Runtime/Library/JsBuiltIn/JsBuiltIn.js [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+                if (this === null || this === undefined) {
+                    throw new TypeError("Array.prototype.values: 'this' is null or undefined");
+                }
+
+                if (typeof (o) !== 'object') {
+                    throw new TypeError("Array Iterator.prototype.next: 'this' is not an Array Iterator object");
+                }
+
+                if (!(o instanceof __chakraLibrary.ArrayIterator)) {
+                    throw new TypeError("Array Iterator.prototype.next: 'this' is not an Array Iterator object");
```
Comment (MikeHolman) [inline lib/Runtime/Library/JsBuiltIn/JsBuiltIn.js]: Can user override TypeError?

File: lib/Runtime/Library/JsBuiltIn/JsBuiltIn.js [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+ 
+    var setPrototype = platform.builtInSetPrototype;
+    var _objectDefineProperty = platform.builtInJavascriptObjectEntryDefineProperty;
+
+    // Object's getter and setter can get overriden on the prototype, in that case while setting the value attributes, we will end up with TypeError
+    // So, we need to set the prototype of attributes to null
+    var ObjectDefineProperty = function (obj, prop, attributes) {
+        _objectDefineProperty(obj, prop, setPrototype(attributes, null));
+    };
+    var CreateObject = platform.builtInJavascriptObjectCreate;
```
Comment (jackhorton) [inline lib/Runtime/Library/JsBuiltIn/JsBuiltIn.js]: Is there a way that we can simplify these aliases for all JsBuiltIns going forward? Like, perhaps we could make the names shorter or have a global standard library written in Javascript that all of these implementation files could use? Something along the lines of https://github.com/Microsoft/ChakraCore/pull/4083/files#diff-fa23cc60499db424efee4a57ac9bafddR105 would be my hope. Having everyone use _.keys or _.defineProperty rather than each files own (hopefully not but possibly) unique alias seems like it would make writing these JsBuiltIns much easier, especially for newer contributors.

File: lib/Runtime/Library/JavascriptLibrary.cpp [hunk 6]
```diff
@@ -18,9 +18,18 @@
...
 #include "Library/ThrowErrorObject.h"
 #include "Library/StackScriptFunction.h"
 
+#include "ByteCode/ByteCodeSerializer.h"
+
+#pragma warning(push)
+#pragma warning(disable:4309) // truncation of constant value
+#pragma warning(disable:4838) // conversion from 'int' to 'const char' requires a narrowing conversion
+
+#pragma warning(pop)
```
Comment (jackhorton) [inline lib/Runtime/Library/JavascriptLibrary.cpp]: Is this intentional? it looks like the warning setup for including the bytecode, but as far as I know this is a no-op right now?

File: lib/Runtime/Library/JsBuiltInEngineInterfaceExtensionObject.cpp [hunk 7]
```diff
@@ -0,0 +1,410 @@
...
+            length = JavascriptConversion::ToLength(JavascriptOperators::OP_GetLength(iterable, scriptContext), scriptContext);
+        }
+
+        return JavascriptNumber::ToVar(length, scriptContext);
+    }
+
+    /*
+    * First parameter is the object onto which prototype should be set; second is the value
+    */
+    Var JsBuiltInEngineInterfaceExtensionObject::EntryJsBuiltIn_Internal_SetPrototype(RecyclableObject *function, CallInfo callInfo, ...)
```
Comment (jackhorton) [inline lib/Runtime/Library/JsBuiltInEngineInterfaceExtensionObject.cpp]: Is there any way we can share native helpers like this between JsBuiltIns and Intl.js? I know there is a similar function in the Intl extension object.

File: lib/Runtime/Library/JsBuiltIn/JsBuiltIn.js [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+                else {
+                    let index = o.nextIndex;
+                    let len = Array.isArray(a) ? a.length : __chakraLibrary.GetArrayLength(a);
+
+                    if (index < len) { // < comparison should happen instead of >= , because len can be NaN
+                        let itemKind = o.kind;
+
+                        o.nextIndex = index + 1;
+
+                        if (itemKind === 0 /*ArrayIterationKind.Key*/) { // TODO (megupta) : Use clean enums here ?
```
Comment (akroshg) [inline lib/Runtime/Library/JsBuiltIn/JsBuiltIn.js]: give .Value more priority over .Key

File: test/DebuggerCommon/parentedDynamicCode3.js.dbg.baseline [hunk 9]
```diff
@@ -7,24 +7,24 @@
...
+        "parentScriptId": 1
       },
       {
-        "scriptId": 4,
-        "parentScriptId": 1,
         "scriptType": "eval code",
         "lineCount": 1,
-        "sourceLength": 11
+        "sourceLength": 11,
+        "scriptId": 5,
```
Comment (akroshg) [inline test/DebuggerCommon/parentedDynamicCode3.js.dbg.baseline]: why this change? Is it due to ChakraLib.js?

File: lib/Runtime/Library/JavascriptFunction.h [hunk 10]
```diff
@@ -42,6 +42,8 @@ namespace Js
         // Need a constructor cache on every function (script and native) to avoid extra checks on the fast path, if the function isn't fixed.
         Field(ConstructorCache*) constructorCache;
 
+        Field(bool) isJsBuiltInCode;
```
Comment (akroshg) [inline lib/Runtime/Library/JavascriptFunction.h]: >isJsBuiltInCode [](start = 20, length = 15)

I don't see this field is initialized anywhere? did you do that somewhere?

File: lib/Runtime/Library/JsBuiltInEngineInterfaceExtensionObject.cpp [hunk 11]
```diff
@@ -0,0 +1,410 @@
...
+        return library->GetUndefined();
+    }
+
+    Var JsBuiltInEngineInterfaceExtensionObject::EntryJsBuiltIn_Internal_GetArrayLengthFunction(RecyclableObject *function, CallInfo callInfo, ...)
+    {
+        EngineInterfaceObject_CommonFunctionProlog(function, callInfo);
+
+        if (callInfo.Count < 2)
+        {
+            return scriptContext->GetLibrary()->GetUndefined();
```
Comment (akroshg) [inline lib/Runtime/Library/JsBuiltInEngineInterfaceExtensionObject.cpp]: We don't expect this one right?, the count will always be greater then 2? Should we assert and return 0 instead of undefined?

File: lib/Runtime/Library/JsBuiltInEngineInterfaceExtensionObject.cpp [hunk 12]
```diff
@@ -0,0 +1,410 @@
...
+        {
+            Js::JavascriptExceptionObject* ex = err.GetAndClear();
+            ex->GetScriptContext();
+            JavascriptError::ThrowTypeError(ex->GetScriptContext(), JSERR_BuiltInNotAvailable);
+        }
+    }
+
+    bool JsBuiltInEngineInterfaceExtensionObject::InitializeJsBuiltInNativeInterfaces(DynamicObject * builtInNativeInterfaces, DeferredTypeHandlerBase * typeHandler, DeferredInitializeMode mode)
+    {
+        typeHandler->Convert(builtInNativeInterfaces, mode, 16);
```
Comment (akroshg) [inline lib/Runtime/Library/JsBuiltInEngineInterfaceExtensionObject.cpp]: >16 [](start = 60, length = 2)

curious why 16?

File: lib/Runtime/Library/JsBuiltInEngineInterfaceExtensionObject.h [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+#if DBG
+        void DumpByteCode() override;
+#endif
+
+        class EntryInfo
+        {
+        public:
+            static NoProfileFunctionInfo JsBuiltIn_RegisterChakraLibraryFunction;
+            static NoProfileFunctionInfo JsBuiltIn_RegisterFunction;
+            static NoProfileFunctionInfo JsBuiltIn_Internal_GetArrayLengthFunction;
```
Comment (akroshg) [inline lib/Runtime/Library/JsBuiltInEngineInterfaceExtensionObject.h]: GetLength instead of GetArrayLength (as we are fetching the 'length' property).

File: lib/Runtime/Library/JsBuiltIn/JsBuiltIn.js (hunk 1 above)
Comment (meg-gupta) [reply lib/Runtime/Library/JsBuiltIn/JsBuiltIn.js]: @akroshg The jit will copy prop the values in either cases. I don't want to add an unconditional write which need not happen.

File: lib/Runtime/Library/JsBuiltIn/JsBuiltIn.js (hunk 2 above)
Comment (meg-gupta) [reply lib/Runtime/Library/JsBuiltIn/JsBuiltIn.js]: Changed to non - enumerable for now.

File: lib/Runtime/Library/JsBuiltIn/JsBuiltIn.js (hunk 8 above)
Comment (meg-gupta) [reply lib/Runtime/Library/JsBuiltIn/JsBuiltIn.js]: Done.

File: test/DebuggerCommon/parentedDynamicCode3.js.dbg.baseline (hunk 9 above)
Comment (meg-gupta) [reply test/DebuggerCommon/parentedDynamicCode3.js.dbg.baseline]: Due to JsBuiltIn.js

File: lib/Runtime/Library/JsBuiltInEngineInterfaceExtensionObject.cpp (hunk 7 above)
Comment (meg-gupta) [reply lib/Runtime/Library/JsBuiltInEngineInterfaceExtensionObject.cpp]: I can try moving it to  EngineInterfaceObject::InitializeCommonNativeInterfaces.

File: lib/Runtime/Library/JsBuiltInEngineInterfaceExtensionObject.h (hunk 13 above)
Comment (meg-gupta) [reply lib/Runtime/Library/JsBuiltInEngineInterfaceExtensionObject.h]: Done.

File: lib/Runtime/Library/JsBuiltInEngineInterfaceExtensionObject.cpp (hunk 12 above)
Comment (meg-gupta) [reply lib/Runtime/Library/JsBuiltInEngineInterfaceExtensionObject.cpp]: Its the initSlotCapacity for the SimpleDictionaryTypeHandler. Changed it to 4, I think that suffices.

File: lib/Runtime/Library/JsBuiltInEngineInterfaceExtensionObject.cpp (hunk 11 above)
Comment (meg-gupta) [reply lib/Runtime/Library/JsBuiltInEngineInterfaceExtensionObject.cpp]: Right, converted to assert.

File: lib/Runtime/Library/JavascriptFunction.h (hunk 10 above)
Comment (meg-gupta) [reply lib/Runtime/Library/JavascriptFunction.h]: This was inherited from the builtins branch. The constructor initialization was missing. It is set in the register* functions and used in a few places. Thanks for catching this!

File: lib/Runtime/Library/JavascriptLibrary.cpp (hunk 6 above)
Comment (meg-gupta) [reply lib/Runtime/Library/JavascriptLibrary.cpp]: Inherited this from the builtins branch and forgot to clean it up! Thanks.

File: lib/Runtime/Library/JsBuiltIn/JsBuiltIn.js (hunk 3 above)
Comment (meg-gupta) [reply lib/Runtime/Library/JsBuiltIn/JsBuiltIn.js]: That'd be bad. Will need an alternative. Thanks.

File: lib/Runtime/Library/JsBuiltIn/JsBuiltIn.js (hunk 4 above)
Comment (meg-gupta) [reply lib/Runtime/Library/JsBuiltIn/JsBuiltIn.js]: Yes. Will need an alternative here. thanks.

**Comment by meg-gupta** [comment]
I have removed the bug fix at https://github.com/Microsoft/ChakraCore/pull/4051 from the PR's commits.

File: lib/Runtime/Library/JsBuiltInEngineInterfaceExtensionObject.cpp [hunk 14]
```diff
@@ -0,0 +1,425 @@
...
+    }
+
+    Var JsBuiltInEngineInterfaceExtensionObject::EntryJsBuiltIn_Internal_InitInternalProperties(RecyclableObject *function, CallInfo callInfo, ...)
+    {
+        EngineInterfaceObject_CommonFunctionProlog(function, callInfo);
+
+        DynamicObject* obj = DynamicObject::FromVar(args.Values[1]);
+        unsigned propCount = TaggedInt::ToUInt32(args.Values[2]);
+        
+        Assert(callInfo.Count == 3 + propCount);
```
Comment (akroshg) [inline lib/Runtime/Library/JsBuiltInEngineInterfaceExtensionObject.cpp]: you don't need this, right? this function can be used for other functions in the future.

Comment (meg-gupta) [reply lib/Runtime/Library/JsBuiltInEngineInterfaceExtensionObject.cpp]: Assert will stay the same for all new functions as well. 3 here is this + object on which we want to set property attributes + count of properties

File: lib/Runtime/Library/JsBuiltIn/JsBuiltIn.js [hunk 15]
```diff
@@ -0,0 +1,132 @@
...
+    // So, we need to set the prototype of attributes to null
+    var ObjectDefineProperty = function (obj, prop, attributes) {
+        _objectDefineProperty(obj, prop, setPrototype(attributes, null));
+    };
+    var CreateObject = platform.builtInJavascriptObjectCreate;
+
+    platform.registerChakraLibraryFunction("ArrayIterator", function (arrayObj, iterationKind) {
+        "use strict";
+        __chakraLibrary.InitInternalProperties(this, 4, "__$arrayObj$__", "__$nextIndex$__", "__$kind$__", "__$internalDone$__");
+        this.__$arrayObj$__ = arrayObj;
```
Comment (akroshg) [inline lib/Runtime/Library/JsBuiltIn/JsBuiltIn.js]: >this.__$arrayObj$__ [](start = 8, length = 19)

make sure you speak with Louis/Ed about this.

Comment (meg-gupta) [reply lib/Runtime/Library/JsBuiltIn/JsBuiltIn.js]: Done. Got approval.

File: lib/Runtime/Library/JsBuiltIn/JsBuiltIn.js [hunk 16]
```diff
@@ -0,0 +1,132 @@
...
+    platform.registerFunction(FunctionsEnum.ArrayEntries, function () {
+        "use strict";
+        if (this === null || this === undefined) {
+            throw new TypeError("Array.prototype.entries: 'this' is null or undefined");
+        }
+        let o = __chakraLibrary.Object(this);
+        return __chakraLibrary.CreateArrayIterator(o, 2 /* ArrayIterationKind.KeyAndValue*/);
+    });
+
+});
```
Comment (dilijev) [inline lib/Runtime/Library/JsBuiltIn/JsBuiltIn.js]: >; [](start = 2, length = 1)

nit: newline at end of file and unnecessary blank line above

File: lib/Runtime/Library/JsBuiltIn/JsBuiltIn.js [hunk 17]
```diff
@@ -0,0 +1,132 @@
...
+        return new __chakraLibrary.ArrayIterator(arrayObj, iterationKind);
+    });
+
+    platform.registerFunction(FunctionsEnum.ArrayKeys, function () {
+        "use strict";
+        if (this === null || this === undefined) {
+            throw new TypeError("Array.prototype.keys: 'this' is null or undefined");
+        }
+        let o = __chakraLibrary.Object(this);
+        return __chakraLibrary.CreateArrayIterator(o, 0 /* ArrayIterationKind.Key*/);
```
Comment (dilijev) [inline lib/Runtime/Library/JsBuiltIn/JsBuiltIn.js]: >0 /* ArrayIterationKind.Key*/ [](start = 54, length = 29)

Can these magic constants be turned into named constants or and/or members of the Enum?

File: lib/Runtime/Library/JsBuiltIn/JsBuiltIn.js [hunk 18]
```diff
@@ -0,0 +1,132 @@
...
+
+                    if (index < len) { // < comparison should happen instead of >= , because len can be NaN
+                        let itemKind = o.__$kind$__;
+
+                        o.__$nextIndex$__ = index + 1;
+
+                        if (itemKind === 1 /*ArrayIterationKind.Value*/) {
+                            value = a[index];
+                        }
+                        else if (itemKind === 0 /*ArrayIterationKind.Key*/) { // TODO (megupta) : Use clean enums here ?
```
Comment (dilijev) [inline lib/Runtime/Library/JsBuiltIn/JsBuiltIn.js]: At this point we probably need to write a JS coding guideline. FWIW the prevailing pattern in library JS is to do `} else if (...) {` on a single line (no newline between `}` and `else`).

File: lib/Runtime/Library/JsBuiltIn/JsBuiltIn.js (hunk 17 above)
Comment (meg-gupta) [reply lib/Runtime/Library/JsBuiltIn/JsBuiltIn.js]: I cannot declare these constants within the global function we use to set up the library. It should be stored on the chakraLib object. Its not neat, but I want to avoid storing constants on the chakraLib object.

File: lib/Runtime/Library/JsBuiltInEngineInterfaceExtensionObject.cpp (hunk 7 above)
Comment (meg-gupta) [reply lib/Runtime/Library/JsBuiltInEngineInterfaceExtensionObject.cpp]: I am going to address this as a separate PR. Also - https://github.com/Microsoft/ChakraCore/pull/3962 is incomplete and requires some special handling for EnginerInterfaceObject's deferred type handler.

File: lib/Runtime/Library/JsBuiltIn/JsBuiltIn.js (hunk 18 above)
Comment (meg-gupta) [reply lib/Runtime/Library/JsBuiltIn/JsBuiltIn.js]: Done. I am squashing CR suggestions with my prior commit, please bear with it!
